File: apps/web/src/v1/candles.ts
import type { MarketCandlesResponse, CandlePrice } from './generated/read-api.ts';

export type CandleIdentity = { marketId: `0x${string}`; memeAsset: string; quoteAsset: string; quoteDecimals: number };
const obj = (v: unknown): v is Record<string, unknown> => typeof v === 'object' && v !== null && !Array.isArray(v);
const uint = (v: unknown): v is string => typeof v === 'string' && /^(0|[1-9][0-9]{0,99})$/.test(v);
const integer = (v: unknown): v is number => Number.isSafeInteger(v) && Number(v) >= 0;
const hash = (v: unknown) => typeof v === 'string' && /^0x[0-9a-f]{64}$/.test(v);
const price = (v: unknown): v is CandlePrice => obj(v) && uint(v.numerator) && uint(v.denominator) && v.numerator !== '0' && v.denominator !== '0';
export const comparePrice = (a: CandlePrice, b: CandlePrice) => {
 const x = BigInt(a.numerator)*BigInt(b.denominator)-BigInt(b.numerator)*BigInt(a.denominator);
 return x<0n?-1:x>0n?1:0;
};
export function validateCandles(v: unknown, chain: number, id: CandleIdentity, from: number, to: number): MarketCandlesResponse {
 const fail = (): never => { throw new Error('Candle data is unavailable or inconsistent'); };
 if (!obj(v) || v.chainId!==chain || v.displayOnly!==true || v.marketId!==id.marketId || v.memeAsset!==id.memeAsset || v.quoteAsset!==id.quoteAsset || v.quoteDecimals!==id.quoteDecimals || v.interval!==3600 || !obj(v.coverage) || !obj(v.series)) return fail();
 const c=v.coverage,s=v.series;
 if (c.from!==from || c.to!==to || !integer(c.anchorNumber) || !integer(c.throughNumber) || !integer(c.projectionNumber) || c.anchorNumber>=c.throughNumber || c.throughNumber>c.projectionNumber || !hash(c.anchorHash) || !hash(c.throughHash) || !hash(c.projectionHash) || s.priceUnit!=='QUOTE_PER_WHOLE_MEME' || s.volumeBasis!=='CURVE_EXCLUDING_FEE_TAX_OR_POOL_CORE' || s.pricePopulation!=='ALL_EXECUTIONS_INCLUDING_INTERNAL_CONVERSIONS' || s.emptyPolicy!=='NULL_OHLC_ZERO_VOLUME' || !Array.isArray(s.candles) || s.candles.length!==(to-from)/3600 || s.candles.length>48) return fail();
 for (const [i,x] of s.candles.entries()) {
  if (!obj(x) || x.timestamp!==from+i*3600 || !integer(x.tradeCount) || !integer(x.internalTradeCount) || !integer(x.unclassifiedTradeCount) || x.tradeCount!==x.internalTradeCount+x.unclassifiedTradeCount) return fail();
  if (!uint(x.memeVolumeRaw) || !uint(x.quoteVolumeRaw) || !uint(x.internalMemeVolumeRaw) || !uint(x.internalQuoteVolumeRaw) || BigInt(x.internalMemeVolumeRaw)>BigInt(x.memeVolumeRaw) || BigInt(x.internalQuoteVolumeRaw)>BigInt(x.quoteVolumeRaw)) return fail();
  if (x.internalTradeCount===0 && (x.internalMemeVolumeRaw!=='0' || x.internalQuoteVolumeRaw!=='0')) return fail();
  if (x.tradeCount===0) { if ([x.open,x.high,x.low,x.close].some(p=>p!==null) || x.memeVolumeRaw!=='0' || x.quoteVolumeRaw!=='0') return fail(); }
  else if (!price(x.open) || !price(x.high) || !price(x.low) || !price(x.close) || x.memeVolumeRaw==='0' || x.quoteVolumeRaw==='0' || comparePrice(x.low,x.open)>0 || comparePrice(x.low,x.close)>0 || comparePrice(x.high,x.open)<0 || comparePrice(x.high,x.close)<0) return fail();
 }
 return v as unknown as MarketCandlesResponse;
}

// Convert only the bounded screen coordinate to Number, never the raw price.
export function candleY(p: CandlePrice, low: CandlePrice, high: CandlePrice): number {
 if (comparePrice(low,high)===0) return 110;
 const n=BigInt(p.numerator)*BigInt(low.denominator)-BigInt(low.numerator)*BigInt(p.denominator);
 const d=BigInt(high.numerator)*BigInt(low.denominator)-BigInt(low.numerator)*BigInt(high.denominator);
 return 210-Number(n*BigInt(high.denominator)*2000n/(d*BigInt(p.denominator)))/10;
}
export function candlePriceLabel(p: CandlePrice): string {
 const n=BigInt(p.numerator),d=BigInt(p.denominator),scaled=n*100000000n/d;
 if (scaled===0n) return '<0.00000001';
 return `${scaled/100000000n}.${(scaled%100000000n).toString().padStart(8,'0')}`.replace(/\.?0+$/,'');
}

File: services/backend-go/internal/analytics/candles.go
package analytics

import (
	"errors"
	"math/big"
	"sort"
	"strconv"
	"strings"
)

var ErrCandles = errors.New("invalid or incomplete candle input")

// CandleTrade is a canonical normalized execution, not a summary/allocation.
// QuoteRaw excludes Curve fee/tax and uses the Pool core delta. Internal
// conversions are included in pool price discovery but broken out in volume.
type CandleTrade struct {
	Source                            CurveSource
	Timestamp                         uint64
	MarketID, MemeAsset, QuoteAsset   string
	QuoteDecimals                     uint8
	MemeRaw, QuoteRaw, Classification string
}

type CandleRange struct {
	ChainID                         uint64
	MarketID, MemeAsset, QuoteAsset string
	QuoteDecimals                   uint8
	From, To, Interval              uint64
}

type CandlePrice struct {
	Numerator   string `json:"numerator"`
	Denominator string `json:"denominator"`
}

type Candle struct {
	Timestamp              uint64       `json:"timestamp"`
	Open                   *CandlePrice `json:"open"`
	High                   *CandlePrice `json:"high"`
	Low                    *CandlePrice `json:"low"`
	Close                  *CandlePrice `json:"close"`
	MemeVolumeRaw          string       `json:"memeVolumeRaw"`
	QuoteVolumeRaw         string       `json:"quoteVolumeRaw"`
	InternalMemeVolumeRaw  string       `json:"internalMemeVolumeRaw"`
	InternalQuoteVolumeRaw string       `json:"internalQuoteVolumeRaw"`
	TradeCount             uint64       `json:"tradeCount"`
	InternalTradeCount     uint64       `json:"internalTradeCount"`
	UnclassifiedTradeCount uint64       `json:"unclassifiedTradeCount"`
}

type CandleSeries struct {
	Candles         []Candle `json:"candles"`
	PriceUnit       string   `json:"priceUnit"`
	VolumeBasis     string   `json:"volumeBasis"`
	PricePopulation string   `json:"pricePopulation"`
	EmptyPolicy     string   `json:"emptyPolicy"`
}

// BuildCandles expects complete input for [From,To). It cannot prove coverage;
// the persistent reader must prove it before exposing zero-volume intervals.
// Empty buckets have null OHLC; no carry-forward or synthetic trade is created.
func BuildCandles(r CandleRange, trades []CandleTrade) (CandleSeries, error) {
	fail := func() (CandleSeries, error) { return CandleSeries{}, ErrCandles }
	intervals := map[uint64]bool{60: true, 300: true, 900: true, 3600: true, 14400: true, 86400: true}
	zeroAddress := "0x" + strings.Repeat("0", 40)
	if r.ChainID == 0 || r.MemeAsset == zeroAddress || (r.QuoteAsset == zeroAddress && r.QuoteDecimals != 18) || !intervals[r.Interval] || r.To <= r.From || r.From%r.Interval != 0 || r.To%r.Interval != 0 || (r.To-r.From)/r.Interval > 2000 || len(trades) > 100000 || !hashRE.MatchString(r.MarketID) || !addressRE.MatchString(r.MemeAsset) || !addressRE.MatchString(r.QuoteAsset) || r.MemeAsset == r.QuoteAsset || r.QuoteDecimals < 6 || r.QuoteDecimals > 18 {
		return fail()
	}
	sorted := append([]CandleTrade{}, trades...)
	heights := map[string]*big.Int{}
	seen := map[string]bool{}
	for _, t := range sorted {
		if !validSource(t.Source, r.ChainID, t.Source.EventKey) || seen[t.Source.EventKey] || t.Timestamp < r.From || t.Timestamp >= r.To || t.MarketID != r.MarketID || t.MemeAsset != r.MemeAsset || t.QuoteAsset != r.QuoteAsset || t.QuoteDecimals != r.QuoteDecimals {
			return fail()
		}
		h, e := uint256(t.Source.BlockNumber)
		if e != nil {
			return fail()
		}
		heights[t.Source.EventKey] = h
		seen[t.Source.EventKey] = true
	}
	sort.Slice(sorted, func(i, j int) bool {
		a, b := sorted[i].Source, sorted[j].Source
		c := heights[a.EventKey].Cmp(heights[b.EventKey])
		if c != 0 {
			return c < 0
		}
		if a.TransactionIndex != b.TransactionIndex {
			return a.TransactionIndex < b.TransactionIndex
		}
		return a.LogIndex < b.LogIndex
	})
	out := CandleSeries{Candles: make([]Candle, (r.To-r.From)/r.Interval), PriceUnit: "QUOTE_PER_WHOLE_MEME", VolumeBasis: "CURVE_EXCLUDING_FEE_TAX_OR_POOL_CORE", PricePopulation: "ALL_EXECUTIONS_INCLUDING_INTERNAL_CONVERSIONS", EmptyPolicy: "NULL_OHLC_ZERO_VOLUME"}
	for i := range out.Candles {
		out.Candles[i] = Candle{Timestamp: r.From + uint64(i)*r.Interval, MemeVolumeRaw: "0", QuoteVolumeRaw: "0", InternalMemeVolumeRaw: "0", InternalQuoteVolumeRaw: "0"}
	}
	add := func(s string, n *big.Int) string { x, _ := new(big.Int).SetString(s, 10); return x.Add(x, n).String() }
	price := func(x *big.Rat) *CandlePrice {
		return &CandlePrice{Numerator: x.Num().String(), Denominator: x.Denom().String()}
	}
	rat := func(x *CandlePrice) *big.Rat {
		n, _ := new(big.Int).SetString(x.Numerator, 10)
		d, _ := new(big.Int).SetString(x.Denominator, 10)
		return new(big.Rat).SetFrac(n, d)
	}
	for i, t := range sorted {
		if i > 0 {
			p := sorted[i-1]
			same := heights[p.Source.EventKey].Cmp(heights[t.Source.EventKey]) == 0
			if t.Timestamp < p.Timestamp || (same && (p.Source.BlockHash != t.Source.BlockHash || p.Timestamp != t.Timestamp || p.Source.LogIndex >= t.Source.LogIndex || (p.Source.TransactionIndex == t.Source.TransactionIndex && p.Source.TransactionHash != t.Source.TransactionHash))) {
				return fail()
			}
		}
		meme, e := uint256(t.MemeRaw)
		if e != nil || meme.Sign() == 0 {
			return fail()
		}
		quote, e := uint256(t.QuoteRaw)
		if e != nil || quote.Sign() == 0 {
			return fail()
		}
		internal := t.Classification == "internal_reward_conversion" || t.Classification == "internal_holder_conversion"
		if !internal && t.Classification != "unclassified" {
			return fail()
		}
		x := new(big.Rat).SetFrac(new(big.Int).Mul(quote, new(big.Int).Exp(big.NewInt(10), big.NewInt(18), nil)), new(big.Int).Mul(meme, new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(r.QuoteDecimals)), nil)))
		c := &out.Candles[(t.Timestamp-r.From)/r.Interval]
		if c.Open == nil {
			c.Open = price(x)
			c.High = price(x)
			c.Low = price(x)
		}
		if x.Cmp(rat(c.High)) > 0 {
			c.High = price(x)
		}
		if x.Cmp(rat(c.Low)) < 0 {
			c.Low = price(x)
		}
		c.Close = price(x)
		c.TradeCount++
		c.MemeVolumeRaw = add(c.MemeVolumeRaw, meme)
		c.QuoteVolumeRaw = add(c.QuoteVolumeRaw, quote)
		if internal {
			c.InternalTradeCount++
			c.InternalMemeVolumeRaw = add(c.InternalMemeVolumeRaw, meme)
			c.InternalQuoteVolumeRaw = add(c.InternalQuoteVolumeRaw, quote)
		} else {
			c.UnclassifiedTradeCount++
		}
	}
	return out, nil
}

func CurveCandleTrade(o CurveObservation) (CandleTrade, error) {
	ts, err := strconv.ParseUint(o.BlockTimestamp, 10, 64)
	if err != nil || strconv.FormatUint(ts, 10) != o.BlockTimestamp {
		return CandleTrade{}, ErrCandles
	}
	return CandleTrade{Source: o.Source, Timestamp: ts, MarketID: o.MarketID, MemeAsset: o.MemeToken, QuoteAsset: o.QuoteAsset, QuoteDecimals: o.QuoteDecimals, MemeRaw: o.Amounts.MemeRaw, QuoteRaw: o.Amounts.QuoteCurveRaw, Classification: o.Classification}, nil
}
func PoolCandleTrade(o PoolObservation) (CandleTrade, error) {
	ts, err := strconv.ParseUint(o.BlockTimestamp, 10, 64)
	if err != nil || strconv.FormatUint(ts, 10) != o.BlockTimestamp {
		return CandleTrade{}, ErrCandles
	}
	return CandleTrade{Source: o.Source, Timestamp: ts, MarketID: o.MarketID, MemeAsset: o.MemeToken, QuoteAsset: o.QuoteAsset, QuoteDecimals: o.QuoteDecimals, MemeRaw: o.Amounts.MemeCoreRaw, QuoteRaw: o.Amounts.QuoteCoreRaw, Classification: o.Classification}, nil
}
